Tasks 1-4, 6-12 and 17

giuliorm opened the pull request.

giuliorm committed:
Task 1 done

giuliorm committed:
Test 1 fixes: FileWriter should write strings with "\r\n" delimiters. BufferedReader and neither FileReader don't work when reading from a file created by FileWriter without string delimiters.

giuliorm committed:
Added \r\n symbols after every string while writing strings to the file. Null checking for every string added. Test testEmptyLinesList added.

giuliorm committed:
Null checking for arguments and for lines added in writeLines method.

giuliorm committed:
Task 2 implementation added

giuliorm committed:
Task 3 implementation added

giuliorm committed:
Task 17 in process

giuliorm committed:
Merge branch 'master' of http://github.com/bmkodoo/javacourses2016 into tests

# Conflicts:
#	src/test/java/com/epam/javacourses2016/task1/Task1Test.java

giuliorm committed:
Merge branch 'master' of http://github.com/bmkodoo/javacourses2016 into tests

# Conflicts:
#	src/test/java/com/epam/javacourses2016/task1/Task1Test.java

giuliorm committed:
Merge branch 'master' of http://github.com/bmkodoo/javacourses2016 into task17

# Conflicts:
#	src/test/java/com/epam/javacourses2016/task1/Task1Test.java

giuliorm committed:
Merge branch 'master' of http://github.com/bmkodoo/javacourses2016 into task17

# Conflicts:
#	src/test/java/com/epam/javacourses2016/task1/Task1Test.java

giuliorm committed:
Merge branch 'master' of http://github.com/bmkodoo/javacourses2016 into ZU

giuliorm committed:
Merge branch 'master' of http://github.com/bmkodoo/javacourses2016 into task17

# Conflicts:
#	src/test/java/com/epam/javacourses2016/task1/Task1Test.java

giuliorm committed:
task 2 changes

giuliorm committed:
Merge branch 'task2' into ZU

giuliorm committed:
Merge branch 'task3' into ZU

giuliorm committed:
Merge branch 'ZU' of http://github.com/giuliorm/javacourses2016 into tests

giuliorm committed:
Merge branch 'ZU' of http://github.com/giuliorm/javacourses2016 into tests

giuliorm committed:
Merge branch 'task17' of http://github.com/giuliorm/javacourses2016 into tests

giuliorm committed:
Task 17 tests added

giuliorm committed:
task 17 done

giuliorm committed:
Merge branch 'task17' into ZU

giuliorm committed:
Merge branch 'ZU' of http://github.com/giuliorm/javacourses2016 into tests

giuliorm committed:
task 17 test added

giuliorm committed:
task 17 test added

giuliorm committed:
task 4 solved

giuliorm committed:
gitignore edited

giuliorm committed:
tasks 6 and 7 solved

giuliorm committed:
task 8 solved

giuliorm committed:
task 9 solved

giuliorm committed:
task 9 fixes

giuliorm committed:
task 10 solved

giuliorm committed:
task 11 test?

giuliorm committed:
task 11

giuliorm committed:
tests enabled

giuliorm committed:
task 1 test merged

giuliorm committed:
task 1 edited

giuliorm committed:
Merge branch 'ZU' of http://github.com/giuliorm/javacourses2016 into task11

giuliorm committed:
Merge branch 'ZU' of http://github.com/giuliorm/javacourses2016 into tests

giuliorm committed:
task 11 tests

giuliorm committed:
Merge branch 'task11' of http://github.com/giuliorm/javacourses2016 into tests

giuliorm committed:
task 11 done

giuliorm committed:
task 11 tests added

giuliorm committed:
Merge branch 'tests' of http://github.com/giuliorm/javacourses2016 into ZU

giuliorm committed:
enable test 12

giuliorm committed:
task 12

giuliorm committed:
Merge branch 'tests' of http://github.com/giuliorm/javacourses2016 into task12

giuliorm committed:
task 12 solved

bmkodoo commented:
Здорово!

Но пожалуйста положи тесты в отдельный PR на мастер
Что следует сделать, что бы вынуть отсюда тесты и оформить их отдельным реквестом:
1. Сделать `git reset --soft master`, все изменения останутся незакомичеными, а мы будем находится на последнем комите в мастере
2. Создать новую ветку `git checkout -b my_tests`
3. Закомитить только тесты, остальное ревёртнуть
4. Ветку my_tests отправить пулл реквестмо на мастер
5. Вернуться в эту ветку, сделать коммит с удалением изменений в тестах (осавить только их включение)

:octocat:

giuliorm commented:
Ага, сделаю

giuliorm commented:
коммит с удалением изменений?.. а разве изменения уже не были закоммичены в п.3? и у меня почему-то получилось больше тестов, чем нужно было

giuliorm commented:
Я вообще все напутала. Как откатиться к предыдущим коммитам?.. как загрузить то, что в репозитории, поверх изменений (revert??). М-да, пытаюсь дальше

giuliorm commented:
Все, разобралась, ура.

giuliorm commented:
Только до сих пор не понятно, что за коммит с удалением измений в тестах и что за включение (enabled=true или имеется в виду "включение" как файлов?)

giuliorm committed:
test enablings

giuliorm committed:
test enablings

giuliorm committed:
Merge remote-tracking branch 'origin/ZU' into ZU

giuliorm committed:
Test 17

giuliorm committed:
Merge branch 'master' of http://github.com/bmkodoo/javacourses2016 into ZU

Conflicts resolved

giuliorm committed:
task 17
Test 17 с сервера абсолютно неправильный, так как там не тестируется результат, который должен выдаваться по условию задачи

giuliorm committed:
task 17

giuliorm committed:
Merge branch 'task12' into ZU

giuliorm commented:
нет, ничего не получается, постоянно откуда-то появляются изменения, которых быть не должно

giuliorm commented:
вроде получше, тперь не знаю, как удалить коммиты с тестами

bmkodoo commented:
[Метод утёнка](https://ru.wikipedia.org/wiki/%D0%9C%D0%B5%D1%82%D0%BE%D0%B4_%D1%83%D1%82%D1%91%D0%BD%D0%BA%D0%B0)

Проще не удалять коммиты, а удалять файлы из ветки. Собственно в моей инструкции это и написано. Главное при этом не забыть сохранить изменения в отдельнйо ветке.

Нужно помнить, что ветке в гите - это всего лишь указатель на коммит. Есть ещё указатель HEAD, это то, где мы находимя сейчас. Есть операция `reset`,  с её помощью можно передвинуть текущий указате HEAD с или без сохранений изменений. Если мы делаем reset --soft, то HEAD передвигается а все изменения остаются как будто незакомичеными, будто мы стоим на том коммите куда переехали reset'ом и проделали эти изменения. Откатившись до мастера, мы таким образом можем отпочковать любую долю изменений из нашей ветки в отдельную ветку. Это и нужно проделать с тестами

bmkodoo commented:
Вижу, ты уже это проделал :+1:

nkarulin commented on src/main/java/com/epam/javacourses2016/task1/SolverTask1.java:
```diff
@@ -16,7 +20,43 @@
+
+    private void writeLines(File output, List<String> lines) {
+        String empty = "";
+        if (output != null && lines != null && lines.size() > 0)
```
If you want to introduce predicate, throw on negative condition
Now you method do nothing with bed input, Java way is to fail when something is wrong

nkarulin commented on src/main/java/com/epam/javacourses2016/task10/SolverTask10.java:
```diff
@@ -17,6 +22,29 @@
-        return null;
+
+        HashMap<String, Integer> englishWords = new HashMap<>();
+        try (FileReader fr = new FileReader(input); BufferedReader br = new BufferedReader(fr)) {
```
`fr` is never used, wrap it in one line like `BufferedReader br = new BufferedReader(new FileReader(input))`

nkarulin commented on src/main/java/com/epam/javacourses2016/task12/SolverTask12.java:
```diff
@@ -9,14 +10,14 @@
+         */
+        List<Integer> transform (List <Integer> integers, int value){
+            integers.sort((item1, item2) -> item1 > item2 ? 1 : item1 < item2 ? -1 : 0);
+            return integers;
```
Are you sure it works?  `value` is never used.

giuliorm replied:
Yeah because:
1. The first case _value_ is not in the array and smaller or bigger than all the values in the array. Then all array will be "in one part" before or after the _value_
2. The _value_ is not in the array, but some elements of the array are bigger than _value_ and some are smaller. Then we have to just to order the array, and if we try to insert the _value_ into the array, some elements will be "before" and "after" the _value_ - it's just a simple ordering.
3. The _value_ is contained in the array - the same situation as in the p. 2

nkarulin replied:
You cheater! Actually, thinking your way the only `integers.sort()` is enough.

nkarulin commented on src/main/java/com/epam/javacourses2016/task17/SolverTask17.java:
```diff
@@ -18,8 +19,57 @@
-    public Set<Point2D> analyze(Set<Segment> segments) {
-        //TODO
-        return null;
+    Set<Point2D> analyze(Set<Segment> segments) {
```
Nice, now it is the fastest solve

What i am talking about: [Пересечение множества отрезков](https://neerc.ifmo.ru/wiki/index.php?title=%D0%9F%D0%B5%D1%80%D0%B5%D1%81%D0%B5%D1%87%D0%B5%D0%BD%D0%B8%D0%B5_%D0%BC%D0%BD%D0%BE%D0%B6%D0%B5%D1%81%D1%82%D0%B2%D0%B0_%D0%BE%D1%82%D1%80%D0%B5%D0%B7%D0%BA%D0%BE%D0%B2), [Алгоритм Бентли-Оттмана](https://neerc.ifmo.ru/wiki/index.php?title=%D0%90%D0%BB%D0%B3%D0%BE%D1%80%D0%B8%D1%82%D0%BC_%D0%91%D0%B5%D0%BD%D1%82%D0%BB%D0%B8-%D0%9E%D1%82%D1%82%D0%BC%D0%B0%D0%BD%D0%B0)

giuliorm replied:
> Nice, now it is the fastest solve

Does that mean my solution is accepted? I'll review the links

nkarulin replied:
Yes, links are just for you to check

giuliorm committed:
Merge with master
Reviews issues fixed.

giuliorm committed:
Reviewed issues corrected.

giuliorm committed:
test corrections

giuliorm committed:
Merge branch 'tests' of https://github.com/giuliorm/javacourses2016 into ZU

# Conflicts:
#	src/test/java/com/epam/javacourses2016/task17/Task17Test.java

giuliorm committed:
Merge branch 'tests' of https://github.com/giuliorm/javacourses2016 into ZU

giuliorm committed:
Tasks errors corrected and tests enabled

giuliorm committed:
Merge branch 'master' of http://github.com/bmkodoo/javacourses2016 into ZU

# Conflicts:
#	src/test/java/com/epam/javacourses2016/task11/SolverTask11Test.java

giuliorm committed:
task 17 test fixes..

bmkodoo commented:
You've got some error with access level. Likely something was missing during conflict resolving.
```
error: analyze(Set<Segment>) is not public in SolverTask17; cannot be accessed from outside package
        Set<Point2D> result = solver.analyze(new HashSet<>(Arrays.asList(segments)));
                                    ^
1 error
:compileTestJava FAILED
```

giuliorm committed:
Merge branch 'master' of http://github.com/bmkodoo/javacourses2016 into ZU

giuliorm committed:
fixes

bmkodoo commented:
master was massively updated

giuliorm commented:
And what now?
Now I have no shippable errors at least

bmkodoo commented:
No. Great job! Continue working
